File: gocode/bank.go
package main

import (
	"fmt"
	"time"
)

const (
	MAX_CLERKS_FOR_BRANCH = 10
)

type Bank interface {
	GetCurrentClerks() []Clerk
	OpenNewClerkDesk() (Clerk, bool)
	DismissClerk(clerk Clerk) bool
}

type bank struct {
	activeClerks []Clerk
	allCustomers []Customer
	security     *BankSecurity
	hiredClerks  int
}

func InitBank() Bank {
	bank := &bank{}
	security := NewBankSecurity(bank)
	bank.security = &security

	newClerk, _ := bank.OpenNewClerkDesk()
	newClerk.Greet(&security)
	return bank
}

func (b *bank) GetCurrentClerks() []Clerk {
	return b.activeClerks
}

func (b *bank) OpenNewClerkDesk() (Clerk, bool) {
	if len(b.activeClerks) >= MAX_CLERKS_FOR_BRANCH {
		return nil, false
	}
	newClerk := NewClerk()
	b.hiredClerks++
	b.activeClerks = append(b.activeClerks, newClerk)
	return newClerk, true
}

func (b *bank) DismissClerk(clerk Clerk) bool {
	if clerk != nil {
		clerk.Dismiss()
		updatedClerks := []Clerk{}
		for _, c := range b.activeClerks {
			if c != clerk {
				updatedClerks = append(updatedClerks, c)
			}
		}
		b.activeClerks = updatedClerks
		return true
	}
	return false
}

func (b *bank) startSpawningCustomers(amount int) {
	fmt.Println("The bank is open!")
	for i := 0; i < amount; i++ {
		fmt.Println("A new customer just arrived")
		cust := NewCustomer()
		b.allCustomers = append(b.allCustomers, cust)
		b.security.HandleNewCustomerArrival(cust)

		// Wait some amount of time (TODO: make it somewhat random)
		time.Sleep(time.Duration(1) * time.Second)
	}

	fmt.Println("The bank is closing~ Waiting for final customer(s) to leave")
	for !b.checkIfAllCustomersDone() {
		fmt.Println("waiting...")
		time.Sleep(time.Duration(1) * time.Second)
	}
	fmt.Println("The bank is closed!")

	happyCustomers := 0
	for _, cus := range b.allCustomers {
		if cus.Mood() == Happy {
			happyCustomers++
		}
	}
	if happyCustomers == amount {
		fmt.Println("Good job handling all customers before they got too upset")
	} else {
		fmt.Printf("Of %d total customers, %d left unhappy. Try improving your clerk assigment some!\n", amount, amount-happyCustomers)
	}
	fmt.Printf("During operating hours you had %d clerks working and ended the day with %d still active\n", b.hiredClerks, len(b.activeClerks))

}

func (b *bank) checkIfAllCustomersDone() bool {
	for _, cus := range b.allCustomers {
		if cus.StillHere() {
			return false
		}
	}
	return true
}

File: core/bank.py
import time
import random
from typing import List, Tuple, Optional

from core.clerk import Clerk, NewClerk
from core.customer import Customer, NewCustomer
from project.security import BankSecurity

MAX_CLERKS_FOR_BRANCH = 6

class Bank:
    def GetCurrentClerks(self) -> List[Clerk]:
        raise NotImplementedError

    def OpenNewClerkDesk(self) -> Tuple[Optional[Clerk], bool]:
        raise NotImplementedError

    def DismissClerk(self, clerk: Clerk) -> bool:
        raise NotImplementedError


# ===== Concrete Bank Implementation =====
class bank(Bank):
    def __init__(self):
        self.activeClerks: List[Clerk] = []
        self.allCustomers: List[Customer] = []
        self.security: Optional[BankSecurity] = None
        self.hiredClerks = 0

    @staticmethod
    def InitBank():
        b = bank()
        security = BankSecurity(b)
        b.security = security

        newClerk = b.OpenNewClerkDesk()
        newClerk.Greet(security)
        return b

    # ---- Interface Methods ----
    def GetCurrentClerks(self) -> List[Clerk]:
        return self.activeClerks

    def OpenNewClerkDesk(self) -> Optional[Clerk]:
        if len(self.activeClerks) >= MAX_CLERKS_FOR_BRANCH:
            return None

        newClerk = NewClerk()
        self.hiredClerks += 1
        self.activeClerks.append(newClerk)
        return newClerk

    def DismissClerk(self, clerk: Clerk) -> bool:
        if clerk is None:
            return False

        clerk.Dismiss()
        self.activeClerks = [c for c in self.activeClerks if c is not clerk]
        return True

    # ---- Main Behavior Methods ----
    def startSpawningCustomers(self, amount: int):
        print("The bank is open!")
        for _ in range(amount):
            print("A new customer just arrived")
            bankCustomer = NewCustomer()
            self.allCustomers.append(bankCustomer)

            self.security.HandleNewCustomerArrival(bankCustomer)

            time.sleep(random.uniform(0.2, 2.0))

        print("The bank is closing~ Waiting for final customer(s) to leave")
        while not self.__checkIfAllCustomersDone():
            print("waiting...")
            time.sleep(1)

        print("The bank is closed!")

        happyCustomers = sum(1 for cus in self.allCustomers if cus.Mood() == "Happy")

        if happyCustomers == amount:
            print("Good job handling all customers before they got too upset")
        else:
            print(
                f"Of {amount} total customers, {amount - happyCustomers} left unhappy. "
                "Try improving your clerk assignment some!"
            )

        print(
            f"During operating hours you had {self.hiredClerks} clerks working and ended "
            f"the day with {len(self.activeClerks)} still active"
        )

    def __checkIfAllCustomersDone(self) -> bool:
        return all(not cus.StillHere() for cus in self.allCustomers)